modules/040-node-manager/hooks/yandex_preemptibly_delete_preemtible_instances.go

```diff
@@ -0,0 +1,224 @@
+		Name:        obj.GetName(),
+		Preemptible: preemptible,
+	}, nil
+}
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10},
+	Queue:        "/modules/cloud-provider-yandex/preemtibly-delete-preemtible-instances",
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:       "machines",
+			ApiVersion: "machine.sapcloud.io/v1alpha1",
+			Kind:       "Machine",
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{"d8-cloud-instance-manager"},
+				},
+			},
+			FilterFunc: applyMachineFilter,
+		},
+		{
+			Name:       "ics",
+			ApiVersion: "deckhouse.io/v1",
+			Kind:       "YandexInstanceClass",
+			FilterFunc: isPreemptibleFilter,
+		},
+	},
+}, deleteMachines)
+
+func deleteMachines(input *go_hook.HookInput) error {
+	timeNow := time.Now().UTC()
+
+	var (
+		icsSnapshot                   = input.Snapshots["ics"]
+		machineSnapshot               = input.Snapshots["machines"]
+		preemptibleInstanceClassesSet = make(map[string]struct{})
+		machines                      []*Machine
+	)
+
+	for _, icRaw := range icsSnapshot {
+		ic, ok := icRaw.(*YandexInstanceClass)
+		if !ok {
+			return fmt.Errorf("failed to assert to *YandexInstanceClass")
+		}
+
+		if ic.Preemptible {
+			preemptibleInstanceClassesSet[ic.Name] = struct{}{}
+		}
+	}
+
+	for _, machineRaw := range machineSnapshot {
+		machine, ok := machineRaw.(*Machine)
+		if !ok {
+			return fmt.Errorf("failed to assert to *Machine")
+		}
+
+		if machine.Terminating {
+			continue
+		}
+
+		if machine.MachineClassKind != "YandexInstanceClass" {
+			continue
+		}
+
+		if _, ok := preemptibleInstanceClassesSet[machine.MachineClassName]; !ok {
+			continue
+		}
+
+		machines = append(machines, machine)
+	}
+
+	if len(machines) == 0 {
+		return nil
+	}
+
+	for _, m := range getMachinesToDelete(timeNow, machines) {
+		input.PatchCollector.Delete("machine.sapcloud.io/v1alpha1", "Machine", "d8-cloud-instance-manager", m)
+	}
+
+	return nil
+}
+
+// delete all after 23h mark
+// afterwards delete in 15 minutes increments, no more than batch size
+func getMachinesToDelete(timeNow time.Time, machines []*Machine) (machinesToDelete []string) {
+	const durationIterations = 12
+	var slidingDuration = preemtibleVMDeletionDuration - time.Hour
+
+	sort.Slice(machines, func(i, j int) bool {
+		return machines[i].CreationTimestamp.Before(&machines[j].CreationTimestamp)
+	})
+
+	batch := len(machines) / durationIterations
+	if batch == 0 {
+		batch = 1
+	}
+
+	var (
+		cursor       int
+		machineCount = len(machines)
+	)
+
+	// short-circuit if there are Machines older than 23 hours
+	for _, m := range machines {
+		if expires(timeNow, m.CreationTimestamp.Time, slidingDuration) {
+			machinesToDelete = append(machinesToDelete, m.Name)
+			cursor++
+		}
+	}
+	if len(machinesToDelete) != 0 {
+		return machinesToDelete
+	}
+
+	for t := 0; t < 12; t++ {
+		slidingDuration -= 15 * time.Minute
```

Comment:
> 15 * time.Minute

NIT: it should also be constant like the preemtibleVMDeletionDuration

Reply:
Done.